// naive-ui/lib/styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.unstablePageHeaderRtl = exports.paginationDark = exports.notificationDark = exports.modalDark = exports.messageDark = exports.menuDark = exports.mentionDark = exports.logDark = exports.loadingBarDark = exports.listDark = exports.layoutDark = exports.unstableInputNumberRtl = exports.inputNumberDark = exports.unstableInputRtl = exports.inputDark = exports.iconDark = exports.gradientTextDark = exports.formDark = exports.emptyDark = exports.elementDark = exports.dynamicTagsDark = exports.dynamicInputDark = exports.dropdownDark = exports.drawerDark = exports.dividerDark = exports.dialogDark = exports.descriptionsDark = exports.datePickerDark = exports.dataTableDark = exports.unstableCollapseRtl = exports.collapseDark = exports.codeDark = exports.unstableCheckboxRtl = exports.checkboxDark = exports.cascaderDark = exports.unstableCardRtl = exports.cardDark = exports.unstableButtonRtl = exports.buttonDark = exports.breadcrumbDark = exports.unstableBadgeRtl = exports.badgeDark = exports.backTopDark = exports.unstableAvatarGroupRtl = exports.avatarDark = exports.autoCompleteDark = exports.anchorDark = exports.alertDark = exports.commonLight = exports.commonDark = void 0;
exports.internalSelectionDark = exports.internalSelectMenuDark = exports.scrollbarDark = exports.watermarkDark = exports.uploadDark = exports.typographyDark = exports.treeSelectDark = exports.treeDark = exports.transferDark = exports.tooltipDark = exports.timelineDark = exports.timePickerDark = exports.thingDark = exports.unstableTagRtl = exports.tagDark = exports.tabsDark = exports.tableDark = exports.switchDark = exports.stepsDark = exports.statisticDark = exports.spinDark = exports.unstableSpaceRtl = exports.spaceDark = exports.sliderDark = exports.selectDark = exports.resultDark = exports.rateDark = exports.unstableRadioRtl = exports.radioDark = exports.progressDark = exports.popselectDark = exports.popoverDark = exports.popconfirmDark = void 0;
var common_1 = require("./_styles/common");
Object.defineProperty(exports, "commonDark", { enumerable: true, get: function () { return common_1.commonDark; } });
Object.defineProperty(exports, "commonLight", { enumerable: true, get: function () { return common_1.commonLight; } });
var styles_1 = require("./alert/styles");
Object.defineProperty(exports, "alertDark", { enumerable: true, get: function () { return styles_1.alertDark; } });
var styles_2 = require("./anchor/styles");
Object.defineProperty(exports, "anchorDark", { enumerable: true, get: function () { return styles_2.anchorDark; } });
var styles_3 = require("./auto-complete/styles");
Object.defineProperty(exports, "autoCompleteDark", { enumerable: true, get: function () { return styles_3.autoCompleteDark; } });
var styles_4 = require("./avatar/styles");
Object.defineProperty(exports, "avatarDark", { enumerable: true, get: function () { return styles_4.avatarDark; } });
var styles_5 = require("./avatar-group/styles");
Object.defineProperty(exports, "unstableAvatarGroupRtl", { enumerable: true, get: function () { return styles_5.avatarGroupRtl; } });
var styles_6 = require("./back-top/styles");
Object.defineProperty(exports, "backTopDark", { enumerable: true, get: function () { return styles_6.backTopDark; } });
var styles_7 = require("./badge/styles");
Object.defineProperty(exports, "badgeDark", { enumerable: true, get: function () { return styles_7.badgeDark; } });
Object.defineProperty(exports, "unstableBadgeRtl", { enumerable: true, get: function () { return styles_7.badgeRtl; } });
var styles_8 = require("./breadcrumb/styles");
Object.defineProperty(exports, "breadcrumbDark", { enumerable: true, get: function () { return styles_8.breadcrumbDark; } });
var styles_9 = require("./button/styles");
Object.defineProperty(exports, "buttonDark", { enumerable: true, get: function () { return styles_9.buttonDark; } });
Object.defineProperty(exports, "unstableButtonRtl", { enumerable: true, get: function () { return styles_9.buttonRtl; } });
var styles_10 = require("./card/styles");
Object.defineProperty(exports, "cardDark", { enumerable: true, get: function () { return styles_10.cardDark; } });
Object.defineProperty(exports, "unstableCardRtl", { enumerable: true, get: function () { return styles_10.cardRtl; } });
var styles_11 = require("./cascader/styles");
Object.defineProperty(exports, "cascaderDark", { enumerable: true, get: function () { return styles_11.cascaderDark; } });
var styles_12 = require("./checkbox/styles");
Object.defineProperty(exports, "checkboxDark", { enumerable: true, get: function () { return styles_12.checkboxDark; } });
Object.defineProperty(exports, "unstableCheckboxRtl", { enumerable: true, get: function () { return styles_12.checkboxRtl; } });
var styles_13 = require("./code/styles");
Object.defineProperty(exports, "codeDark", { enumerable: true, get: function () { return styles_13.codeDark; } });
var styles_14 = require("./collapse/styles");
Object.defineProperty(exports, "collapseDark", { enumerable: true, get: function () { return styles_14.collapseDark; } });
Object.defineProperty(exports, "unstableCollapseRtl", { enumerable: true, get: function () { return styles_14.collapseRtl; } });
var styles_15 = require("./data-table/styles");
Object.defineProperty(exports, "dataTableDark", { enumerable: true, get: function () { return styles_15.dataTableDark; } });
var styles_16 = require("./date-picker/styles");
Object.defineProperty(exports, "datePickerDark", { enumerable: true, get: function () { return styles_16.datePickerDark; } });
var styles_17 = require("./descriptions/styles");
Object.defineProperty(exports, "descriptionsDark", { enumerable: true, get: function () { return styles_17.descriptionsDark; } });
var styles_18 = require("./dialog/styles");
Object.defineProperty(exports, "dialogDark", { enumerable: true, get: function () { return styles_18.dialogDark; } });
var styles_19 = require("./divider/styles");
Object.defineProperty(exports, "dividerDark", { enumerable: true, get: function () { return styles_19.dividerDark; } });
var styles_20 = require("./drawer/styles");
Object.defineProperty(exports, "drawerDark", { enumerable: true, get: function () { return styles_20.drawerDark; } });
var styles_21 = require("./dropdown/styles");
Object.defineProperty(exports, "dropdownDark", { enumerable: true, get: function () { return styles_21.dropdownDark; } });
var styles_22 = require("./dynamic-input/styles");
Object.defineProperty(exports, "dynamicInputDark", { enumerable: true, get: function () { return styles_22.dynamicInputDark; } });
var styles_23 = require("./dynamic-tags/styles");
Object.defineProperty(exports, "dynamicTagsDark", { enumerable: true, get: function () { return styles_23.dynamicTagsDark; } });
var styles_24 = require("./element/styles");
Object.defineProperty(exports, "elementDark", { enumerable: true, get: function () { return styles_24.elementDark; } });
var styles_25 = require("./empty/styles");
Object.defineProperty(exports, "emptyDark", { enumerable: true, get: function () { return styles_25.emptyDark; } });
var styles_26 = require("./form/styles");
Object.defineProperty(exports, "formDark", { enumerable: true, get: function () { return styles_26.formDark; } });
var styles_27 = require("./gradient-text/styles");
Object.defineProperty(exports, "gradientTextDark", { enumerable: true, get: function () { return styles_27.gradientTextDark; } });
var styles_28 = require("./icon/styles");
Object.defineProperty(exports, "iconDark", { enumerable: true, get: function () { return styles_28.iconDark; } });
var styles_29 = require("./input/styles");
Object.defineProperty(exports, "inputDark", { enumerable: true, get: function () { return styles_29.inputDark; } });
Object.defineProperty(exports, "unstableInputRtl", { enumerable: true, get: function () { return styles_29.inputRtl; } });
var styles_30 = require("./input-number/styles");
Object.defineProperty(exports, "inputNumberDark", { enumerable: true, get: function () { return styles_30.inputNumberDark; } });
Object.defineProperty(exports, "unstableInputNumberRtl", { enumerable: true, get: function () { return styles_30.inputNumberRtl; } });
var styles_31 = require("./layout/styles");
Object.defineProperty(exports, "layoutDark", { enumerable: true, get: function () { return styles_31.layoutDark; } });
var styles_32 = require("./list/styles");
Object.defineProperty(exports, "listDark", { enumerable: true, get: function () { return styles_32.listDark; } });
var styles_33 = require("./loading-bar/styles");
Object.defineProperty(exports, "loadingBarDark", { enumerable: true, get: function () { return styles_33.loadingBarDark; } });
var styles_34 = require("./log/styles");
Object.defineProperty(exports, "logDark", { enumerable: true, get: function () { return styles_34.logDark; } });
var styles_35 = require("./mention/styles");
Object.defineProperty(exports, "mentionDark", { enumerable: true, get: function () { return styles_35.mentionDark; } });
var styles_36 = require("./menu/styles");
Object.defineProperty(exports, "menuDark", { enumerable: true, get: function () { return styles_36.menuDark; } });
var styles_37 = require("./message/styles");
Object.defineProperty(exports, "messageDark", { enumerable: true, get: function () { return styles_37.messageDark; } });
var styles_38 = require("./modal/styles");
Object.defineProperty(exports, "modalDark", { enumerable: true, get: function () { return styles_38.modalDark; } });
var styles_39 = require("./notification/styles");
Object.defineProperty(exports, "notificationDark", { enumerable: true, get: function () { return styles_39.notificationDark; } });
var styles_40 = require("./pagination/styles");
Object.defineProperty(exports, "paginationDark", { enumerable: true, get: function () { return styles_40.paginationDark; } });
var styles_41 = require("./page-header/styles");
Object.defineProperty(exports, "unstablePageHeaderRtl", { enumerable: true, get: function () { return styles_41.pageHeaderRtl; } });
var styles_42 = require("./popconfirm/styles");
Object.defineProperty(exports, "popconfirmDark", { enumerable: true, get: function () { return styles_42.popconfirmDark; } });
var styles_43 = require("./popover/styles");
Object.defineProperty(exports, "popoverDark", { enumerable: true, get: function () { return styles_43.popoverDark; } });
var styles_44 = require("./popselect/styles");
Object.defineProperty(exports, "popselectDark", { enumerable: true, get: function () { return styles_44.popselectDark; } });
var styles_45 = require("./progress/styles");
Object.defineProperty(exports, "progressDark", { enumerable: true, get: function () { return styles_45.progressDark; } });
var styles_46 = require("./radio/styles");
Object.defineProperty(exports, "radioDark", { enumerable: true, get: function () { return styles_46.radioDark; } });
Object.defineProperty(exports, "unstableRadioRtl", { enumerable: true, get: function () { return styles_46.radioRtl; } });
var styles_47 = require("./rate/styles");
Object.defineProperty(exports, "rateDark", { enumerable: true, get: function () { return styles_47.rateDark; } });
var styles_48 = require("./result/styles");
Object.defineProperty(exports, "resultDark", { enumerable: true, get: function () { return styles_48.resultDark; } });
var styles_49 = require("./select/styles");
Object.defineProperty(exports, "selectDark", { enumerable: true, get: function () { return styles_49.selectDark; } });
var styles_50 = require("./slider/styles");
Object.defineProperty(exports, "sliderDark", { enumerable: true, get: function () { return styles_50.sliderDark; } });
var styles_51 = require("./space/styles");
Object.defineProperty(exports, "spaceDark", { enumerable: true, get: function () { return styles_51.spaceDark; } });
Object.defineProperty(exports, "unstableSpaceRtl", { enumerable: true, get: function () { return styles_51.spaceRtl; } });
var styles_52 = require("./spin/styles");
Object.defineProperty(exports, "spinDark", { enumerable: true, get: function () { return styles_52.spinDark; } });
var styles_53 = require("./statistic/styles");
Object.defineProperty(exports, "statisticDark", { enumerable: true, get: function () { return styles_53.statisticDark; } });
var styles_54 = require("./steps/styles");
Object.defineProperty(exports, "stepsDark", { enumerable: true, get: function () { return styles_54.stepsDark; } });
var styles_55 = require("./switch/styles");
Object.defineProperty(exports, "switchDark", { enumerable: true, get: function () { return styles_55.switchDark; } });
var styles_56 = require("./table/styles");
Object.defineProperty(exports, "tableDark", { enumerable: true, get: function () { return styles_56.tableDark; } });
var styles_57 = require("./tabs/styles");
Object.defineProperty(exports, "tabsDark", { enumerable: true, get: function () { return styles_57.tabsDark; } });
var styles_58 = require("./tag/styles");
Object.defineProperty(exports, "tagDark", { enumerable: true, get: function () { return styles_58.tagDark; } });
Object.defineProperty(exports, "unstableTagRtl", { enumerable: true, get: function () { return styles_58.tagRtl; } });
var styles_59 = require("./thing/styles");
Object.defineProperty(exports, "thingDark", { enumerable: true, get: function () { return styles_59.thingDark; } });
var styles_60 = require("./time-picker/styles");
Object.defineProperty(exports, "timePickerDark", { enumerable: true, get: function () { return styles_60.timePickerDark; } });
var styles_61 = require("./timeline/styles");
Object.defineProperty(exports, "timelineDark", { enumerable: true, get: function () { return styles_61.timelineDark; } });
var styles_62 = require("./tooltip/styles");
Object.defineProperty(exports, "tooltipDark", { enumerable: true, get: function () { return styles_62.tooltipDark; } });
var styles_63 = require("./transfer/styles");
Object.defineProperty(exports, "transferDark", { enumerable: true, get: function () { return styles_63.transferDark; } });
var styles_64 = require("./tree/styles");
Object.defineProperty(exports, "treeDark", { enumerable: true, get: function () { return styles_64.treeDark; } });
var styles_65 = require("./tree-select/styles");
Object.defineProperty(exports, "treeSelectDark", { enumerable: true, get: function () { return styles_65.treeSelectDark; } });
var styles_66 = require("./typography/styles");
Object.defineProperty(exports, "typographyDark", { enumerable: true, get: function () { return styles_66.typographyDark; } });
var styles_67 = require("./upload/styles");
Object.defineProperty(exports, "uploadDark", { enumerable: true, get: function () { return styles_67.uploadDark; } });
var styles_68 = require("./watermark/styles");
Object.defineProperty(exports, "watermarkDark", { enumerable: true, get: function () { return styles_68.watermarkDark; } });
// danger zone, internal styles
var styles_69 = require("./_internal/scrollbar/styles");
Object.defineProperty(exports, "scrollbarDark", { enumerable: true, get: function () { return styles_69.scrollbarDark; } });
var styles_70 = require("./_internal/select-menu/styles");
Object.defineProperty(exports, "internalSelectMenuDark", { enumerable: true, get: function () { return styles_70.internalSelectMenuDark; } });
var styles_71 = require("./_internal/selection/styles");
Object.defineProperty(exports, "internalSelectionDark", { enumerable: true, get: function () { return styles_71.internalSelectionDark; } });
